### PR TITLE
Adding a caching system as well as the base structure of the xp cog.

### PR DESCRIPTION
**Pull request by itsVale**
I implemented a caching system for bulk storing items in a `collections.defaultdict` that assigns a list to every key and appends items to it. Using that, I started to write the XP system that also has most important parts implemented.

### CODE REVIEW

File: bot/utils/cache.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    return frozenset(args)
+
+
+default_key = functools.partial(functools._make_key, typed=False)
+typed_key = functools.partial(functools._make_key, typed=True)
+
+
+# Here's the original: https://github.com/Rapptz/RoboDanny/blob/rewrite/cogs/utils/cache.py
+# Modified to allow custom key args and the strategy is determined by the max_size
+def cache(max_size=128, make_key=default_key):
```
Comment (ef4203) [inline bot/utils/cache.py]: Cache is too complex, you are using ``def`` statements inside a ``def statement``.

File: bot/utils/cache.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    def decorator(func):
+        if max_size is None:
+            cache = collections.defaultdict(list)
+            get_stats = lambda: (0, 0)
```
Comment (ef4203) [inline bot/utils/cache.py]: Do not assign lambda expression, use a ``def``.

File: bot/cogs/xp_system.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import math
+
+from ..utils import cache
+
+
+class XP:
+    def __init__(self, bot):
+        self.bot = bot
+        self.factor = (1 / 1.2)
```
Comment (ef4203) [inline bot/cogs/xp_system.py]: Since this is a constant, move this out of the constructor.

File: bot/utils/cache.py (hunk 1 above)
Comment (itsVale) [reply bot/utils/cache.py]: This basically is the way how decorators work in Python. Things like
```python
def wrap_and_store(key, coro):
    async def function():
        value = await coro()
        cache[key].append(value)
        return value

    return function()
```
are style things that become useful when awaiting the decorated function.

Comment (ef4203) [reply bot/utils/cache.py]: Sure, add a ``# noqa`` at the end of the line, so the pylava/pylama check passes.

File: bot/cogs/xp_system.py (hunk 3 above)
Comment (itsVale) [reply bot/cogs/xp_system.py]: Done.

File: bot/utils/cache.py (hunk 2 above)
Comment (itsVale) [reply bot/utils/cache.py]: Done.

**Pull request merged**